File: src/client/ZWBF/components/UI/ZWBFTabManager.ts
import { getText } from "@asledgehammer/pipewrench";
import { ISCharacterInfoWindow, ISWindow, ISLayoutManager } from "@asledgehammer/pipewrench/client";

interface TabInfo {
	viewName: string;
	ui: any; // You can make this more specific based on your UI components
}
interface Layout {
	tabs?: string;
	current?: string;
	[key: string]: any;
}
export class ZWBFTabManager {
	private static tabs: Record<string, TabInfo> = {};
	private static methodsOverridden = false;

	/**
	 * Adds a new tab to the character info window
	 * @param tabName The name for the tab
	 * @param ui UI component to show
	 */
	static addTab(tabName: string, ui: any): void {
		const viewName = `${tabName}View`;

		// Store tab information
		ZWBFTabManager.tabs[tabName] = {
			viewName,
			ui
		};

		// Override methods only once
		if (!ZWBFTabManager.methodsOverridden) {
			ZWBFTabManager.overrideISCharacterInfoWindowMethods();
			ZWBFTabManager.methodsOverridden = true;
		}
	}

	/**
	 * Overrides necessary methods in ISCharacterInfoWindow
	 */
	private static overrideISCharacterInfoWindowMethods(): void {
		// Store original methods
		const originalCreateChildren = ISCharacterInfoWindow.prototype.createChildren;
		const originalOnTabTornOff = ISCharacterInfoWindow.prototype.onTabTornOff;
		const originalPrerender = ISCharacterInfoWindow.prototype.prerender;
		const originalSaveLayout = ISCharacterInfoWindow.prototype.SaveLayout;

		// Extend createChildren
		ISCharacterInfoWindow.prototype.createChildren = function () {
			originalCreateChildren.call(this);

			for (const [tabName, tabInfo] of Object.entries(ZWBFTabManager.tabs)) {
				const viewName = tabInfo.viewName;
				const ui = tabInfo.ui;

				(this as any)[viewName] = ui;
				(this as any)[viewName].setPositionPixel(0, 0);
				(this as any)[viewName].infoText = getText(`UI_${tabName}_Info`);
				(this as any)[viewName].closeButton.setVisible(false);

				// Prevent the tab content from being dragged
				(this as any)[viewName].onMouseDown = () => {
					(this as any)[viewName].setX(0);
					(this as any)[viewName].setY((ISWindow as any).TitleBarHeight);
				};

				this.panel.addView(getText(`UI_${tabName}`), (this as any)[viewName]);
			}
		};

		// Extend onTabTornOff
		ISCharacterInfoWindow.prototype.onTabTornOff = function (view: any, window: any) {
			for (const [tabName, tabInfo] of Object.entries(ZWBFTabManager.tabs)) {
				if (this.playerNum === 0 && view === (this as any)[tabInfo.viewName]) {
					ISLayoutManager.RegisterWindow(`charinfowindow.${tabName}`, ISCharacterInfoWindow, window);
				}
			}
			originalOnTabTornOff.call(this, view, window);
		};

		// Extend prerender
		ISCharacterInfoWindow.prototype.prerender = function () {
			originalPrerender.call(this);

			for (const tabInfo of Object.values(ZWBFTabManager.tabs)) {
				const viewName = tabInfo.viewName;
				if ((this as any)[viewName] === this.panel.getActiveView()) {
					this.setWidth((this as any)[viewName].getWidth());
					this.setHeight(((ISWindow as any).TitleBarHeight * 2) + (this as any)[viewName].getHeight());
				}
			}
		};

		// Extend SaveLayout
		ISCharacterInfoWindow.prototype.SaveLayout = function (name: string, layout: Layout) {
			originalSaveLayout.call(this, name, layout);

			for (const [tabName, tabInfo] of Object.entries(ZWBFTabManager.tabs)) {
				const subSelf = (this as any)[tabInfo.viewName];
				if (subSelf && subSelf.parent === this.panel) {
					if (!layout.tabs) {
						layout.tabs = tabName;
					} else {
						layout.tabs = `${layout.tabs},${tabName}`;
					}
					if (subSelf === this.panel.getActiveView()) {
						layout.current = tabName;
					}
				}
			}
		};
	}
}

File: src/client/ZWBF/components/UI/ZWBFTabManager.lua
--- Localized global functions from PZ
local getText = getText

--- @class ISCharacterInfoWindow
--- @field panel table 
--- @field playerNum number 
--- @field setWidth function
--- @field setHeight function
local ISCharacterInfoWindow = ISCharacterInfoWindow

--- @class ISWindow
--- @field TitleBarHeight number
local ISWindow = ISWindow

--- @class ISLayoutManager
--- @field RegisterWindow function
local ISLayoutManager = ISLayoutManager


--- ZWBF Character Info Tab Manager Class
--- This class manages the tabs in the character info window.
--- It allows adding new tabs and handles their display.
--- @class ZWBFTabManager
--- @field tabs table A table to store the tabs
local ZWBFTabManager = {}
ZWBFTabManager.__index = ZWBFTabManager

--- Static table to store tabs
ZWBFTabManager.tabs = {}

--- Constructor
function ZWBFTabManager:new()
    local instance = setmetatable({}, self)
    return instance
end

--- Adds a new tab
--- @param tabName string The name for the tab
--- @param ui any UI component to show
function ZWBFTabManager:addTab(tabName, ui)
    local viewName = tabName .. "View"

    -- Store tab information
    ZWBFTabManager.tabs[tabName] = {
        viewName = viewName,
        ui = ui,
    }

    -- Override methods only once
    if not ZWBFTabManager.methodsOverridden then
        self:overrideISCharacterInfoWindowMethods()
        ZWBFTabManager.methodsOverridden = true
    end
end

--- Overrides necessary methods in ISCharacterInfoWindow
function ZWBFTabManager:overrideISCharacterInfoWindowMethods()
    local originalCreateChildren = ISCharacterInfoWindow.createChildren
    local originalOnTabTornOff = ISCharacterInfoWindow.onTabTornOff
    local originalPrerender = ISCharacterInfoWindow.prerender
    local originalSaveLayout = ISCharacterInfoWindow.SaveLayout

    -- Extend createChildren
    function ISCharacterInfoWindow:createChildren()
        originalCreateChildren(self)

        for tabName, tabInfo in pairs(ZWBFTabManager.tabs) do
            local viewName = tabInfo.viewName
            local ui = tabInfo.ui

            self[viewName] = ui
            self[viewName]:setPositionPixel(0, 0)
            self[viewName].infoText = getText("UI_" .. tabName .. "_Info")
            self[viewName].closeButton:setVisible(false)

            -- Prevent the tab content from being dragged
            self[viewName].onMouseDown = function()
                self[viewName]:setX(0)
                self[viewName]:setY(ISWindow.TitleBarHeight)
            end

            self.panel:addView(getText("UI_" .. tabName), self[viewName])
        end
    end

    -- Extend onTabTornOff
    function ISCharacterInfoWindow:onTabTornOff(view, window)
        for tabName, tabInfo in pairs(ZWBFTabManager.tabs) do
            if self.playerNum == 0 and view == self[tabInfo.viewName] then
                ISLayoutManager.RegisterWindow('charinfowindow.' .. tabName, ISCharacterInfoWindow, window)
            end
        end
        originalOnTabTornOff(self, view, window)
    end

    -- Extend prerender
    function ISCharacterInfoWindow:prerender()
        originalPrerender(self)
        for _, tabInfo in pairs(ZWBFTabManager.tabs) do
            local viewName = tabInfo.viewName
            if self[viewName] == self.panel:getActiveView() then
                self:setWidth(self[viewName]:getWidth())
                self:setHeight((ISWindow.TitleBarHeight * 2) + self[viewName]:getHeight())
            end
        end
    end

    -- Extend SaveLayout
    function ISCharacterInfoWindow:SaveLayout(name, layout)
        originalSaveLayout(self, name, layout)

        for tabName, tabInfo in pairs(ZWBFTabManager.tabs) do
            local subSelf = self[tabInfo.viewName]
            if subSelf and subSelf.parent == self.panel then
                if not layout.tabs then
                    layout.tabs = tabName
                else
                    layout.tabs = layout.tabs .. ',' .. tabName
                end
                if subSelf == self.panel:getActiveView() then
                    layout.current = tabName
                end
            end
        end
    end
end

local module = { ZWBFTabManager = ZWBFTabManager }
return module;
